tests/SafearrayTests.cpp

```diff
@@ -0,0 +1,1574 @@
+                                                , wil::unique_dispatch_safearray            \
+                                                , wil::unique_variant_safearray             \
+                                            )
+
+#define RUN_TEST(test)                      WI_FOREACH(test                                 \
+                                                , char                                      \
+                                                , short                                     \
+                                                , long                                      \
+                                                , int                                       \
+                                                , long long                                 \
+                                                , unsigned char                             \
+                                                , unsigned short                            \
+                                                , unsigned long                             \
+                                                , unsigned int                              \
+                                                , unsigned long long                        \
+                                                , float                                     \
+                                                , double                                    \
+                                                , VARIANT_BOOL                              \
+                                                , DATE                                      \
+                                                , BSTR                                      \
+                                                , LPUNKNOWN                                 \
+                                                , LPDISPATCH                                \
+                                                , VARIANT                                   \
+                                            )
+
+#define RUN_TEST_SPECIAL(test)              WI_FOREACH(test                                 \
+                                                , BSTR                                      \
+                                                , LPUNKNOWN                                 \
+                                                , LPDISPATCH                                \
+                                                , VARIANT                                   \
+                                            )
+
+
+// {5D80EC64-6694-4F49-B0B9-CCAA65467D12}
+//static const GUID IID_IAmForTesting =
+//{ 0x5d80ec64, 0x6694, 0x4f49, { 0xb0, 0xb9, 0xcc, 0xaa, 0x65, 0x46, 0x7d, 0x12 } };
+
+interface __declspec(uuid("5D80EC64-6694-4F49-B0B9-CCAA65467D12"))
+    IAmForTesting : public IUnknown
+{
+    STDMETHOD(GetID)(LONG* pOut) = 0;
+};
+
+class TestComObject : virtual public IAmForTesting,

```

Comment:
can you use `winrt::implements` here instead to avoid the IUnknown boilerplate? use WINRT_CUSTOM_MODULE_LOCK, get_module_lock to enable global object counting.

Reply:
Not at all familiar with WinRT but am willing to learn.  It will just take a little while to make that change.

Reply:
I converted the class to use WinRT for it's COM implementation and then realized that `winrt::implements` requires C++17 for the features that it uses, in which case, I'd rather keep it as-is.

Reply:
Unless there's no requirement to support pre-C++17 anymore... Then sure.